sklearn/tree/_criterion.pyx

```diff
@@ -574,20 +575,26 @@ cdef class ClassificationCriterion(Criterion):
             memcpy(dest, &self.sum_total[k, 0], self.n_classes[k] * sizeof(float64_t))
             dest += self.max_n_classes
 
-    cdef void clip_node_value(self, float64_t * dest, float64_t lower_bound, float64_t upper_bound) noexcept nogil:
-        """Clip the value in dest between lower_bound and upper_bound for monotonic constraints.
-
+    cdef void clip_node_value(
+        self, float64_t * dest, float64_t lower_bound, float64_t upper_bound
+    ) noexcept nogil:
+        """Clip the values in dest such that predicted probabilities stay between
+        `lower_bound` and `upper_bound` when monotonic constraints are enforced.
         Note that monotonicity constraints are only supported for:
         - single-output trees and
         - binary classifications.
         """
-        if dest[0] < lower_bound:
-            dest[0] = lower_bound
-        elif dest[0] > upper_bound:
-            dest[0] = upper_bound
-
-        # Class proportions for binary classification must sum to 1.
-        dest[1] = 1 - dest[0]
+        cdef:
+            float64_t total_weighted_count = dest[0] + dest[1]
+            float64_t scaled_lower_bound = (<float32_t>lower_bound) * total_weighted_count
+            float64_t scaled_upper_bound = (<float32_t>upper_bound) * total_weighted_count
+        if dest[0] < scaled_lower_bound:
+            dest[0] = scaled_lower_bound
+        elif dest[0] > scaled_upper_bound:
+            dest[0] = scaled_upper_bound
+
+        # Values for binary classification must sum to total_weighted_count.
+        dest[1] = total_weighted_count - dest[0]
```

Comment:
```suggestion
        dest[1] = max(0, total_weighted_count - dest[0])
```
Do we need to protect `dest[1]` to be non-negative?

Reply:
`lower_bound` and `upper_bound` are computed as positive class frequencies (or averages of positive class frequencies), so they should always be in the `[0, 1]` range, and thus `dest[1]` should always be non-negative.

If we decide to protect as you suggest though, note that we should make sure the invariant `dest[0] + dest[1] = total_weighted_count` holds.